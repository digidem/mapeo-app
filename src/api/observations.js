// @flow
import fetch from 'fetch';
import { jsonRequest } from './base';

export interface CreateRequest {
  lat: number;
  lon: number;
}

export type UpdateRequest = {
  id: string
};

class Observation {
  static list = () =>
    jsonRequest({
      method: 'GET',
      route: '/observations'
    });

  static create = (observation: CreateRequest) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: {
        device_id: '1',
        ...observation
      }
    });

  static update = (observation: UpdateRequest) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: {
        ...observation
      }
    });
}

export default Observation;
