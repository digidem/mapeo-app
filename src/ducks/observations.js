// @flow

import update from 'immutability-helper';
import {keyBy} from 'lodash';
import {create} from '../lib/redux';
import {createSelector} from '../lib/selector';
import type {StoreState} from '../types/redux';
import type {Observation} from '../types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const observations = keyBy(action.payload, 'id');

    if (state.observations) {
      Object.keys(state.observations).forEach(id => observations[id] = {
        ...observations[id],
        ...state.observations[id],
      });
    }

    const newState = update(state, {observations: {$set: observations}});

    return newState;
  }
});

export const {
  type: OBSERVATION_CREATE,
  action: observationCreate,
  reducer: observationCreateReducer
} = create('OBSERVATION_CREATE', {
  start: (state, action) => {
    const observation = action.meta;

    if (!observation.lat && !observation.lon && state.gps.data) {
      observation.lat = state.gps.data.latitude;
      observation.lon = state.gps.data.longitude;
    }

    const newState = update(state, {selectedObservation: {$set: observation}});

    return newState;
  },
  success: (state, action) => {
    const newState = update(state, {selectObservation: {$set: action.payload}});

    return newState;
  }
});

export const {
  type: OBSERVATION_UPDATE,
  action: observationUpdate,
  reducer: observationUpdateReducer
} = create('OBSERVATION_UPDATE', {
  start: (state, action) => {
    let newState;
    if (state.selectedObservation &&
        action.meta.id === state.selectedObservation.id) {
      newState = update(state, {
        selectedObservation: {
          $set: {...state.selectedObservation, ...action.meta},
        }
      });
    } else {
      newState = update(state, {
        observations: {
          [action.meta.id]: {
            $set: {...state.observations[action.meta.id], ...action.meta}
          }
        }
      });
    }

    return newState;
  }
});

export const {
  type: OBSERVATION_SELECT,
  action: observationSelect,
  reducer: observationSelectReducer
} = create('OBSERVATION_SELECT', {
  start: (state, action) => {
    if (action.meta === undefined && state.selectedObservation) {
      return update(state, {
        selectedObservation: {$set: action.meta},
        observations: {
          [state.selectedObservation.id]: {
            $set: {
              ...(state.observations[state.selectedObservation.id] || {}),
              ...state.selectedObservation,
            }
          }
        }
      });
    } else {
      return update(state, {selectedObservation: {$set: action.meta}});
    }
  }
});

export const selectObservation = createSelector(
    [(state: StoreState, id: string): Observation =>
         state.app.observations[id]],
    (observation: Observation): Observation => observation,
    (observation: Observation, id: string): string => id);

export default [
  observationListReducer, observationCreateReducer, observationUpdateReducer,
  observationSelectReducer
];
