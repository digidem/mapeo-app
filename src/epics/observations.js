// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import {
  OBSERVATION_LIST,
  observationList,
  OBSERVATION_CREATE,
  observationCreate,
  OBSERVATION_UPDATE,
  observationUpdate
} from '../ducks/observations';
import { Action } from '../types/redux';
import type { Observation as ObservationType } from '../types/observation';
import type { StoreState } from '../types/redux';
import Observation from '../api/observations';
import type { CreateRequest, UpdateRequest } from '../api/observations';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, ObservationType[]>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() =>
      Observation.list().map(observations => {
        console.log('RN - Observations List', observations);
        return observationList('', observations);
      })
    );

export const observationCreateEpic = (
  action$: ActionsObservable<Action<CreateRequest, ObservationType>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_CREATE)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Observation.create(action.meta).map(observation => {
        console.log('RN - Observation Created');
        return observationCreate(action.meta, observation);
      })
    );

// export const observationUpdateEpic = (
//   action$: ActionsObservable<Action<UpdateRequest, ObservationType>>,
//   store: StoreState
// ) =>
//   action$
//     .ofType(OBSERVATION_UPDATE)
//     .filter(action => action.status === 'Start')
//     .flatMap(action =>
//       Observation.update(action.meta).map(observation => {
//         console.log('RN - Observation Updated - ', action.meta, observation);
//         return observationUpdate(action.meta, observation);
//       })
//     );

export default [
  observationListEpic,
  observationCreateEpic
  // observationUpdateEpic
];
