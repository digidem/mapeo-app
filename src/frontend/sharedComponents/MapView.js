// @flow
import React from "react";
import { View, StyleSheet } from "react-native";
import MapboxGL from "@react-native-mapbox-gl/maps";
import debug from "debug";

// import type { MapStyle } from "../types";
import AddButton from "./AddButton";
import { LocationFollowingIcon, LocationNoFollowIcon } from "./icons";
import IconButton from "./IconButton";
import withNavigationFocus from "../lib/withNavigationFocus";
import type {
  LocationContextType,
  PositionType
} from "../context/LocationContext";
import type { ObservationsMap } from "../context/ObservationsContext";

const log = debug("mapeo:MapView");

const mapboxStyles = {
  observation: {
    circleColor: "#F29D4B",
    circleRadius: 5,
    circleStrokeColor: "#fff",
    circleStrokeWidth: 2
  }
};

type ObservationFeature = {
  type: "Feature",
  geometry: {
    type: "Point",
    coordinates: [number, number] | [number, number, number]
  },
  properties: {| id: string |}
};

/**
 * Convert a map of observations into a GeoJSON FeatureCollection
 *
 * @param {{ [string]: ObservationType }} Observations
 * @returns GeoJSON FeatureCollection with Features that have the observation
 * location and id
 */
function mapObservationsToFeatures(
  obsMap: ObservationsMap
): ObservationFeature[] {
  const features = [];
  for (const obs of obsMap.values()) {
    // Only include observations with a location in the map view
    if (
      typeof obs.value.lon === "number" &&
      typeof obs.value.lat === "number"
    ) {
      features.push({
        type: "Feature",
        geometry: {
          type: "Point",
          coordinates: [obs.value.lon, obs.value.lat]
        },
        properties: {
          id: obs.id
        }
      });
    }
  }
  return features;
}

class ObservationMapLayer extends React.PureComponent<{
  onPress: Function,
  observations: ObservationsMap
}> {
  render() {
    const { onPress, observations } = this.props;
    const featureCollection = {
      type: "FeatureCollection",
      features: mapObservationsToFeatures(observations)
    };
    return (
      <MapboxGL.ShapeSource
        onPress={onPress}
        id={`observations-source`}
        shape={featureCollection}
      >
        <MapboxGL.CircleLayer id={`circles`} style={mapboxStyles.observation} />
      </MapboxGL.ShapeSource>
    );
  }
}

type Props = {
  observations: ObservationsMap,
  styleURL: string,
  location: LocationContextType,
  onAddPress: () => any,
  onPressObservation: (observationId: string) => any,
  isFocused: boolean
};

type State = {
  // True if the map is following user location
  following: boolean
};

class MapView extends React.Component<Props, State> {
  static defaultProps = {
    onAddPress: () => {},
    onPressObservation: () => {}
  };
  currentZoom: number;
  map: any;
  initialPosition: void | null | PositionType;

  constructor(props: Props) {
    super(props);
    MapboxGL.setAccessToken(
      "pk.eyJ1IjoiZ21hY2xlbm5hbiIsImEiOiJSaWVtd2lRIn0.ASYMZE2HhwkAw4Vt7SavEg"
    );
    MapboxGL.setTelemetryEnabled(false);
    log("accessToken set");
    this.initialPosition =
      props.location.position || props.location.savedPosition;
    this.state = {
      following:
        !!props.location.provider &&
        props.location.provider.locationServicesEnabled
    };
  }

  // We only use the location prop (which contains the app GPS location) for the
  // first render of the map. After that location updates come from the native
  // map view, so we don't want to re-render this component every time there is
  // a GPS update, only when the location provider status changes, which we use
  // to render the map in follow-mode or not.
  shouldComponentUpdate(nextProps: Props, nextState: State) {
    return (
      shallowDiffers(this.props, nextProps, ["location"]) ||
      shallowDiffers(this.state, nextState) ||
      shallowDiffers(this.props.location.provider, nextProps.location.provider)
    );
  }

  handleObservationPress = (e: {
    nativeEvent?: {
      payload?: {
        properties?: { id: string }
      }
    }
  }) => {
    const pressedFeature = e.nativeEvent && e.nativeEvent.payload;
    if (!pressedFeature || !pressedFeature.properties) return;

    const observationId = pressedFeature.properties.id;
    log("handle obs press", observationId);
    const { observations, onPressObservation } = this.props;
    if (observations.get(observationId)) {
      onPressObservation(observationId);
    } else {
      log(
        "Warning: pressed feature with id '" +
          observationId +
          "' but could not find matching observation"
      );
    }
  };

  handleMapViewRef = (c: any) => {
    this.map = c;
  };

  handleRegionDidChange = (e: any) => {
    if (!e.properties.isUserInteraction) return;
    // Any user interaction with the map switches follow mode to false
    this.setState({ following: false });
    this.currentZoom = e.properties.zoomLevel;
    console.log(e.properties);
  };

  getFollowZoomLevel() {
    // If the user has zoomed the map, we keep to that zoom level when we switch
    // to "follow mode"
    return this.currentZoom || 12;
  }

  handleLocationPress = () => {
    const { location } = this.props;
    if (!(location.provider && location.provider.locationServicesEnabled))
      return;
    this.setState(state => ({ following: !state.following }));
  };

  render() {
    const {
      observations,
      onAddPress,
      styleURL,
      isFocused,
      location
    } = this.props;
    const initialCoords = this.initialPosition
      ? [
          this.initialPosition.coords.longitude,
          this.initialPosition.coords.latitude
        ]
      : [0, 0];
    const initialZoom = this.initialPosition ? 8 : 0;
    const locationServicesEnabled =
      location.provider && location.provider.locationServicesEnabled;
    return (
      <>
        <MapboxGL.MapView
          style={{ flex: 1 }}
          ref={this.handleMapViewRef}
          maxZoomLevel={22}
          logoEnabled={false}
          pitchEnabled={false}
          rotateEnabled={false}
          onPress={this.handleObservationPress}
          compassEnabled={false}
          styleURL={styleURL}
          onRegionDidChange={this.handleRegionDidChange}
        >
          {locationServicesEnabled && (
            <MapboxGL.UserLocation visible={isFocused} />
          )}
          <MapboxGL.Camera
            centerCoordinate={initialCoords}
            zoomLevel={initialZoom}
            followUserLocation={isFocused && this.state.following}
            followUserMode="normal"
            followZoomLevel={this.getFollowZoomLevel()}
            animationMode="flyTo"
            triggerKey={this.state.following}
          />
          <ObservationMapLayer
            onPress={this.handleObservationPress}
            observations={observations}
          />
        </MapboxGL.MapView>
        <AddButton onPress={onAddPress} />
        <View style={styles.locationButton}>
          <IconButton onPress={this.handleLocationPress}>
            {this.state.following ? (
              <LocationFollowingIcon />
            ) : (
              <LocationNoFollowIcon />
            )}
          </IconButton>
        </View>
      </>
    );
  }
}

export default withNavigationFocus(MapView);

// Shallow compare objects, but omitting certain keys from the comparison
function shallowDiffers(a: any, b: any, omit: string[] = []) {
  for (const i in a) if (!(i in b)) return true;
  for (const i in b) {
    if (a[i] !== b[i] && omit.indexOf(i) === -1) return true;
  }
  return false;
}

const styles = StyleSheet.create({
  locationButton: {
    position: "absolute",
    right: 20,
    bottom: 20
  }
});
