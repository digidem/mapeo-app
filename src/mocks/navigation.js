import type { NavigationScreenProp } from 'react-navigation';

export const createNavigationScreenProp = () => {
  return {
    state: {},
    dispatch: jest.fn(),
    addListener: jest.fn(),
    getParam: jest.fn(),
    isFocused: jest.fn(),
    goBack: jest.fn(),
    navigate: jest.fn(),
    setParams: jest.fn()
  };
};
