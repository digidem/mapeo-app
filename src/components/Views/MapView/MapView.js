// @flow
import React from 'react';
import {
  Image,
  ImageBackground,
  StyleSheet,
  View,
  TouchableOpacity,
  Dimensions,
  PermissionsAndroid
} from 'react-native';
import { withNavigationFocus } from 'react-navigation';
import MapboxGL from '@mapbox/react-native-mapbox-gl';
import Icon from 'react-native-vector-icons/MaterialIcons';
import type { Observation } from '@types/observation';
import { isEmpty, size, map, filter } from 'lodash';
import type { CreateRequest, UpdateRequest } from '@api/observations';
import env from '../../../../env.json';

import AddButton from '../../../images/add-button.png';
import Gradient from '../../../images/gradient-overlay.png';
import { WHITE } from '../../../lib/styles';
import { applyObservationDefaults } from '../../../models/observations';
import type { Resource } from '../../../types/redux';
import type { GPSState } from '../../../types/gps';

export type StateProps = {
  observations: {
    [id: string]: Observation
  },
  selectedObservation?: Observation,
  gps?: GPSState
};

export type DispatchProps = {
  listObservations: () => void,
  createObservation: (observation: CreateRequest) => void,
  updateObservation: (observation: UpdateRequest) => void,
  goToCategories: () => void,
  goToObservationDetail: () => void,
  selectObservation: (observation: Observation) => void
};

type Props = {
  isFocused: boolean
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center',
    backgroundColor: 'white'
  },

  mapPlaceholder: {
    backgroundColor: '#000',
    color: '#FFF',
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },

  map: {
    flex: 1,
    alignSelf: 'stretch'
  },

  newObservation: {
    alignSelf: 'center'
  },

  buttons: {
    flexDirection: 'row'
  },

  btn: {
    backgroundColor: '#4444FF',
    padding: 10,
    margin: 10
  },

  btnText: {
    color: 'white'
  },

  info: {
    textAlign: 'left',
    color: '#333333',
    marginBottom: 5
  }
});

const mapboxStyles = MapboxGL.StyleSheet.create({
  point: {
    circleColor: '#5086EC',
    circleRadius: 5,
    circleStrokeColor: '#fff',
    circleStrokeWidth: 2
  },
  observation: {
    circleColor: '#F29D4B',
    circleRadius: 5,
    circleStrokeColor: '#fff',
    circleStrokeWidth: 2
  },
  currentLocation: {
    circleColor: '#2657B1',
    circleRadius: 7,
    circleStrokeColor: '#fff',
    circleStrokeWidth: 3
  }
});

class MapView extends React.Component<Props & StateProps & DispatchProps> {
  constructor() {
    super();

    MapboxGL.setAccessToken(env.accessToken);
  }

  componentDidMount() {
    const { observations, listObservations } = this.props;

    listObservations();
  }

  shouldComponentUpdate(nextProps: Props & StateProps & DispatchProps) {
    if (nextProps.isFocused) {
      return nextProps !== this.props;
    }

    return false;
  }

  map: any;

  handleCreateObservation = () => {
    const {
      createObservation,
      observations,
      updateObservation,
      goToCategories,
      selectedObservation
    } = this.props;
    const initialObservation = applyObservationDefaults({
      id: size(observations) + 1
    });
    goToCategories();

    createObservation(initialObservation);
  };

  handleObservationPress = (id: string) => {
    const {
      observations,
      selectObservation,
      goToObservationDetail
    } = this.props;

    if (observations[id]) {
      selectObservation(observations[id]);
      goToObservationDetail();
    }
  };

  handleLongPress = (point: Object) => {
    const { coordinates } = point.geometry;
    const { createObservation, observations, goToCategories } = this.props;
    const initialObservation = applyObservationDefaults({
      id: size(observations) + 1,
      lat: coordinates[1],
      lon: coordinates[0]
    });

    goToCategories();
    createObservation(initialObservation);
  };

  handleMapViewRef = c => {
    this.map = c;
  };

  goToCurrentLocation = () => {
    const { gps } = this.props;

    if (this.map && gps) {
      this.map.moveTo([gps.longitude, gps.latitude], 200);
    }
  };

  render() {
    const { observations, gps } = this.props;

    return (
      <View
        style={{
          width: Dimensions.get('window').width,
          height: Dimensions.get('window').height
        }}
      >
        <View style={{ flex: 1 }}>
          <MapboxGL.MapView
            style={{ flex: 1 }}
            centerCoordinate={gps ? [gps.longitude, gps.latitude] : undefined}
            ref={this.handleMapViewRef}
            zoomLevel={12}
            logoEnabled
            onLongPress={this.handleLongPress}
            compassEnabled={false}
            textureMode
          >
            {!!observations && !isEmpty(observations)
              ? map(observations, (o: Observation) => (
                <MapboxGL.ShapeSource
                  onPress={() => this.handleObservationPress(o.id)}
                  key={o.id}
                  id={`observations-${o.id}`}
                  shape={{
                      type: 'Feature',
                      geometry: {
                        type: 'Point',
                        coordinates: [o.lon, o.lat]
                      },
                      properties: {
                        name: o.name
                      }
                    }}
                >
                  <MapboxGL.CircleLayer
                    id={`circles-${o.id}`}
                    style={mapboxStyles.observation}
                  />
                </MapboxGL.ShapeSource>
                ))
              : null}
            {gps && (
              <MapboxGL.ShapeSource
                key="current-location"
                id="current-location"
                shape={{
                  type: 'Feature',
                  geometry: {
                    type: 'Point',
                    coordinates: [gps.longitude, gps.latitude]
                  }
                }}
              >
                <MapboxGL.CircleLayer
                  id="circles-current-location"
                  style={mapboxStyles.currentLocation}
                />
              </MapboxGL.ShapeSource>
            )}
          </MapboxGL.MapView>
          <View
            style={{
              position: 'absolute',
              bottom: 25,
              alignItems: 'center',
              backgroundColor: 'transparent',
              zIndex: 5,
              flexDirection: 'row',
              justifyContent: 'space-around',
              width: Dimensions.get('window').width
            }}
          >
            <View style={{ width: 50 }} />
            <TouchableOpacity
              onPress={this.handleCreateObservation}
              style={styles.newObservation}
            >
              <Image
                source={AddButton}
                style={{
                  width: 125,
                  height: 125
                }}
              />
            </TouchableOpacity>
            <TouchableOpacity onPress={this.goToCurrentLocation}>
              <Icon color={WHITE} name="my-location" size={50} />
            </TouchableOpacity>
          </View>
        </View>
        <View
          style={{
            position: 'absolute',
            left: 0,
            backgroundColor: 'transparent'
          }}
        >
          <ImageBackground
            source={Gradient}
            style={{
              width: Dimensions.get('window').width,
              height: 100
            }}
          />
        </View>
      </View>
    );
  }
}

export default withNavigationFocus(MapView);
