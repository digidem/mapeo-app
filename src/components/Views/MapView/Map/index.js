// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import {
  observationCreate,
  observationUpdate,
  observationSelect
} from '../../../../ducks/observations';
import { observationSource } from '../../../../ducks/observationSource';
import { StoreState } from '../../../../types/redux';
import Map from './Map';
import type { DispatchProps, StateProps } from './Map';

const mapStateToProps = (state: StoreState) => ({
  observations: state.observations,
  selectedObservation: state.selectedObservation,
  coords: state.gps.coords,
  selectedStyle: state.map.selectedStyle
});

function mapDispatchToProps(dispatch: Dispatch<*>) {
  return {
    createObservation: observation => dispatch(observationCreate(observation)),
    updateObservation: observation => dispatch(observationUpdate(observation)),
    selectObservation: o => dispatch(observationSelect(o)),
    updateObservationSource: () => dispatch(observationSource('map'))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Map);
