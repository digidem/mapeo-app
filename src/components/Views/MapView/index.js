// @flow
import { connect } from 'react-redux';
import MapView from './MapView';
import { drawerClose, drawerOpen } from '../../../ducks/drawers';
import { observationList } from '../../../ducks/observations';

const mapStateToProps = state => ({
  showSavedModal: state.modals.saved
});

const mapDispatchToProps = (dispatch: Dispatch<*>) => ({
  onDrawerClose: () => dispatch(drawerClose('observations')),
  onDrawerOpen: () => dispatch(drawerOpen('observations')),
  listObservations: () => dispatch(observationList(''))
});

export default connect(mapStateToProps, mapDispatchToProps)(MapView);
