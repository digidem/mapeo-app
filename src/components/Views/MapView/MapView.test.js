// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import MapView from './MapView';
import { createObservation } from '../../../mocks/observations';
import { createNavigationScreenProp } from '../../../mocks/navigation';

jest.mock('../ObservationsView', () => () => null);
jest.mock('./Map', () => () => null);
jest.mock('../../Base/SavedModal', () => () => null);
jest.mock('backoff-rxjs', () => ({ retryBackoff: () => null }));

describe('MapView tests', () => {
  test('snapshots', () => {
    let tree;
    tree = renderer.create(
      <MapView
        navigation={createNavigationScreenProp()}
        onDrawerOpen={jest.fn()}
        onDrawerClose={jest.fn()}
        listObservations={jest.fn()}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
