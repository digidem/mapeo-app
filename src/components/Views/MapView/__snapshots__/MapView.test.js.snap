// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView tests snapshots 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withNavigation(withNavigationFocus(MapView))
    navigation={
      Object {
        "addListener": [Function],
        "isFocused": [Function],
      }
    }
    selectedObservation={
      Object {
        "categoryId": 19,
        "created": "2018-03-25T07:30:00.000Z",
        "icon": Object {
          "testUri": "../../../src/images/categories/category_18.png",
        },
        "id": 1,
        "lat": 50.5,
        "link": "linkID",
        "lon": 50.5,
        "media": Array [],
        "name": "Oil Spill",
        "notes": "This is pretty bad. It smells pretty noxious, and the entire pond is affected.",
        "observedBy": "You",
        "type": "Observation",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": Object {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_calculateChangedBits": null,
      "_changedBits": 0,
      "_currentRenderer": null,
      "_currentValue": undefined,
      "_defaultValue": undefined,
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_currentRenderer": null,
        "_currentValue": undefined,
        "_defaultValue": undefined,
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
